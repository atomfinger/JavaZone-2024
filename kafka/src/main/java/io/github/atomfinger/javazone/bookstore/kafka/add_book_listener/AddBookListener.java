package io.github.atomfinger.javazone.bookstore.kafka.add_book_listener;

import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.stereotype.Component;

import io.github.atomfinger.javazone.bookstore.service.BookService;

@Component
public class AddBookListener {

    private final BookService bookService;

    public AddBookListener(BookService bookService) {
        this.bookService = bookService;
    }

    @KafkaListener(groupId = "BookstoreService", topics = "bookstore.cmd.add-book.1", containerFactory = "addBookKafkaListenerContainerFactory")
    public void listen(AddBookMessage message) {
        bookService.addBook(message.toBook());
    }
}
