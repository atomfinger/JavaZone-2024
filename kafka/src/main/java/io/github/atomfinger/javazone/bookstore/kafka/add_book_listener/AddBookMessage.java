package io.github.atomfinger.javazone.bookstore.kafka.add_book_listener;

import io.github.atomfinger.javazone.bookstore.persistence.entities.Book;

import java.util.Date;

public record AddBookMessage(Long bookId, String title, String description, String isbn, String authorName,
        Integer pageNumbers, Date publishedDate, String genre) {

    public Book toBook() {
        var book = new Book();
        book.setBookId(bookId);
        book.setTitle(title);
        book.setDescription(description);
        book.setIsbn(isbn);
        book.setAuthorName(authorName);
        book.setPageNumbers(pageNumbers);
        book.setPublishedDate(publishedDate);
        book.setGenre(genre);
        return book;
    }

}
