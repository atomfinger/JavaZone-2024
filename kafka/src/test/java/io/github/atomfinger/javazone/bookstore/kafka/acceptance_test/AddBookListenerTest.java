package io.github.atomfinger.javazone.bookstore.kafka.acceptance_test;

import io.github.atomfinger.javazone.bookstore.kafka.add_book_listener.AddBookMessage;
import org.approvaltests.JsonApprovals;
import org.junit.jupiter.api.Test;

import java.sql.Date;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

class AddBookListenerTest extends AcceptanceTestBase {

    @Test
    public void given_that_we_add_a_new_book_then_new_book_should_be_added_to_db() {
        var input = new AddBookMessage(
                1L,
                "Effective Java",
                "A comprehensive guide to programming in Java.",
                "9780134685991",
                "Joshua Bloch",
                416,
                Date.valueOf("2018-01-06"),
                "Programming"
        );
        kafkaTemplate().send("bookstore.cmd.add-book.1", "key", input);
        await().atMost(5, SECONDS).until(() -> bookRepository.count() > 0);
        var result = bookRepository.findAll().iterator().next();
        JsonApprovals.verifyAsJson(result);
    }
}