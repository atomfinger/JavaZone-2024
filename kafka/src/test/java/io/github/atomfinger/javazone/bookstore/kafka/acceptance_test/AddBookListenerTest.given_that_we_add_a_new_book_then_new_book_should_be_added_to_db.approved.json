{
  "bookId": 1,
  "title": "Effective Java",
  "description": "A comprehensive guide to programming in Java.",
  "isbn": "9780134685991",
  "authorName": "Joshua Bloch",
  "pageNumbers": 416,
  "publishedDate": "Jan 6, 2018",
  "genre": "Programming"
}